src/valarray.hpp

```diff
@@ -153,4 +191,101 @@ template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto min(q
 	return min_val;
 }
 
+// max(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto max(quokka::valarray<T, d> const &v) -> T
+{
+  static_assert(d >= 1);
+  T max_val = v[0]; // v must have at least 1 element
+
+  for (size_t i = 0; i < v.size(); ++i) {
+    max_val = std::max(max_val, v[i]);
+  }
+  return max_val;
+}
+
+// sum(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto sum(quokka::valarray<T, d> const &v) -> T
+{
+  T sum_val = 0;
+  for (size_t i = 0; i < v.size(); ++i) {
+    sum_val += v[i];
+  }
+  return sum_val;
+}
+
+// hasnan(array)
+template <typename T, int d> AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto hasnan(quokka::valarray<T, d> const &v) -> bool
+{
+  for (size_t i = 0; i < v.size(); ++i) {
+    if (std::isnan(v[i])) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// array > array
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator>(quokka::valarray<T, d> const &a, quokka::valarray<T, d> const &b) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] > b[i];
+  }
+  return comp;
+}
+
+// array > scalar
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator>(quokka::valarray<T, d> const &a, T const &scalar) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] > scalar;
+  }
+  return comp;
+}
+
+// array < array
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator<(quokka::valarray<T, d> const &a, quokka::valarray<T, d> const &b) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] < b[i];
+  }
+  return comp;
+}
+
+// array < scalar
+template <typename T, int d>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator<(quokka::valarray<T, d> const &a, T const &scalar) -> quokka::valarray<bool, d>
+{
+  quokka::valarray<bool, d> comp;
+  for (size_t i = 0; i < a.size(); ++i) {
+    comp[i] = a[i] < scalar;
+  }
+  return comp;
+}
+
+// define fillin(valarray) for quokka::valarray
+template <typename T, int d> 
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void fillin(quokka::valarray<T, d> &v, T const &scalar)
+{
+  for (size_t i = 0; i < v.size(); ++i) {
+    v[i] = scalar;
+  }
+}
+
+// // do slicing with valarray, e.g. array_positive_terms = array[array > 0.]. Not working
+// template <typename T, int d>
+// AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto operator[](quokka::valarray<T, d> const &v, quokka::valarray<bool, d> const &mask) -> quokka::valarray<T, d>
+// {
+//   quokka::valarray<T, d> sliced;
+//   for (size_t i = 0; i < v.size(); ++i) {
+//     sliced[i] = mask[i] ? v[i] : 0.;
+//   }
+//   return sliced;
+// }
```

Comment:
Can you remove the commented-out code? This can be kept on a separate branch until it's working and you want to merge it. In general, we want to avoid commented-out code in the main branch (there are some places where it exists and should be removed, but we haven't gotten around to it).